embassy-stm32/src/sdmmc/mod.rs

```diff
@@ -550,6 +563,44 @@ impl<'d, T: Instance, Dma: SdmmcDma<T>> Sdmmc<'d, T, Dma> {
         regs.data_interrupts(false);
         state.wake();
     }
+
+    /// Returns kernel clock (SDIOCLK) for the SD-card facing domain
+    fn kernel_clock() -> Hertz {
+        cfg_if::cfg_if! {
+            // TODO, these could not be implemented, because required clocks are not exposed in RCC:
+            // - H7 uses pll1_q_ck or pll2_r_ck depending on SDMMCSEL
+            // - L1 uses pll48
+            // - L4 uses clk48(pll48)
+            // - L4+, L5, U5 uses clk48(pll48) or PLLSAI3CLK(PLLP) depending on SDMMCSEL
+            if #[cfg(stm32f1)] {
+                // F1 uses AHB1(HCLK), which is correct in PAC
+                T::frequency()
+            } else if #[cfg(any(stm32f2, stm32f4))] {
+                // F2, F4 always use pll48
+                critical_section::with(|_| unsafe {
+                    crate::rcc::get_freqs().pll48
+                }).expect("PLL48 is required for SDIO")
+            } else if #[cfg(stm32f7)] {
+                critical_section::with(|_| unsafe {
+                    use core::any::TypeId;
+                    let sdmmcsel = if TypeId::of::<T>() == TypeId::of::<crate::peripherals::SDMMC1>() {
+                        crate::pac::RCC.dckcfgr2().read().sdmmc1sel()
+                    } else {
+                        crate::pac::RCC.dckcfgr2().read().sdmmc2sel()
+                    };
```

Comment:
Is there a better way to express this?

Reply:
make the `kernel_clock()` method part of the `trait Instance`, implement it inside the `foreach_peripheral!`.

To make it generate different code for sdmmc1/2 without having to copypaste everything, you could do this:

```rust
macro_rules! kernel_clock {
    (sdmmc1) => {
        // code here
    };
    (sdmmc2) => {
        // code here
    };
}

foreach_peripheral!(
    (sdmmc, $inst:ident) => {
        impl sealed::Instance for peripherals::$inst {
            ...
            fn kernel_clock() -> Hertz {
                kernel_clock!($inst);
            }
        }
    };
);
```

Super ugly, but I think it's less bad than using `TypeId`.